Example run for https://github.com/librenms/librenms/pull/19409 
Level 6: 25 errors (100% type hinting problems)
Level 7: 30 errors (A couple of `expects string, string|false given`)
Level 8: 30 errors (exact same as level 7)
```
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   app/Casts/EncryptedArray.php
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------
  10     Class App\Casts\EncryptedArray implements generic interface Illuminate\Contracts\Database\Eloquent\CastsAttributes but does not specify its types: TGet, TSet
         🪪  missingType.generics
  17     Method App\Casts\EncryptedArray::get() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ ----------------------------------------------------------------------------------------------------------------------------
  Line   app/Data/DeviceSecretRepository.php
 ------ ----------------------------------------------------------------------------------------------------------------------------
  63     Method App\Data\DeviceSecretRepository::getRawSecretData() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ ----------------------------------------------------------------------------------------------------------------------------

 ------ --------------------------------------------------------------------------------------------------------------------------
  Line   app/Data/Secrets/IpmiSecret.php
 ------ --------------------------------------------------------------------------------------------------------------------------
  38     Method App\Data\Secrets\IpmiSecret::fromArray() has parameter $data with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  48     Method App\Data\Secrets\IpmiSecret::rules() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  58     Method App\Data\Secrets\IpmiSecret::getUiSchema() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ --------------------------------------------------------------------------------------------------------------------------

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------
  Line   app/Data/Secrets/SecretData.php
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------
  32     Class App\Data\Secrets\SecretData implements generic interface Illuminate\Contracts\Support\Arrayable but does not specify its types: TKey, TValue
         🪪  missingType.generics
  37     Method App\Data\Secrets\SecretData::fromArray() has parameter $data with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  42     Method App\Data\Secrets\SecretData::rules() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  47     Method App\Data\Secrets\SecretData::getUiSchema() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  59     Method App\Data\Secrets\SecretData::toJson() should return string but returns string|false.
         🪪  return.type
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------

 ------ --------------------------------------------------------------------------------------------------------------------------
  Line   app/Data/Secrets/SnmpSecret.php
 ------ --------------------------------------------------------------------------------------------------------------------------
  89     Method App\Data\Secrets\SnmpSecret::fromArray() has parameter $data with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  104    Method App\Data\Secrets\SnmpSecret::rules() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  118    Method App\Data\Secrets\SnmpSecret::getUiSchema() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ --------------------------------------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   app/Http/Controllers/SecretController.php
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------
  150    Method App\Http\Controllers\SecretController::maskPasswordFields() has parameter $data with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  150    Method App\Http\Controllers\SecretController::maskPasswordFields() has parameter $schema with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  150    Method App\Http\Controllers\SecretController::maskPasswordFields() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  161    Method App\Http\Controllers\SecretController::restoreMaskedFields() has parameter $newData with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  161    Method App\Http\Controllers\SecretController::restoreMaskedFields() has parameter $originalData with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  161    Method App\Http\Controllers\SecretController::restoreMaskedFields() has parameter $schema with no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  161    Method App\Http\Controllers\SecretController::restoreMaskedFields() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ ---------------------------------------------------------------------------------------------------------------------------------------
  Line   app/Models/Secret.php
 ------ ---------------------------------------------------------------------------------------------------------------------------------------
  12     Class App\Models\Secret uses generic trait Illuminate\Database\Eloquent\Factories\HasFactory but does not specify its types: TFactory
         🪪  missingType.generics
 ------ ---------------------------------------------------------------------------------------------------------------------------------------

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   database/factories/SecretFactory.php
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  8      Class Database\Factories\SecretFactory extends generic class Illuminate\Database\Eloquent\Factories\Factory but does not specify its types: TModel
         🪪  missingType.generics
  15     PHPDoc type string of property Database\Factories\SecretFactory::$model is not covariant with PHPDoc type class-string<Illuminate\Database\Eloquent\Model> of overridden property Illuminate\Database\Eloquent\Factories\Factory<Illuminate\Database\Eloquent\Model>::$model.
         🪪  property.phpDocType
         💡  You can fix 3rd party PHPDoc types with stub files:
         💡  https://phpstan.org/user-guide/stub-files
  22     Method Database\Factories\SecretFactory::definition() return type has no value type specified in iterable type array.
         🪪  missingType.iterableValue
         💡  See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  22     Return type (array) of method Database\Factories\SecretFactory::definition() should be compatible with return type (array<model property of Illuminate\Database\Eloquent\Model, mixed>) of method Illuminate\Database\Eloquent\Factories\Factory<Illuminate\Database\Eloquent\Model>::definition()
         🪪  method.childReturnType
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ --------------------------------------------------------------------------
  Line   database/migrations/2026_04_01_162956_migrate_snmp_secrets.php
 ------ --------------------------------------------------------------------------
  31     Parameter #1 $string of function md5 expects string, string|false given.
         🪪  argument.type
  44     Parameter #1 $string of function md5 expects string, string|false given.
         🪪  argument.type
  72     Parameter #1 $string of function md5 expects string, string|false given.
         🪪  argument.type
  96     Parameter #1 $string of function md5 expects string, string|false given.
         🪪  argument.type
 ------ --------------------------------------------------------------------------

 [ERROR] Found 30 errors
```

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -53,6 +53,7 @@ jobs:
       - name: Checkout Code
         uses: actions/checkout@v5
         with:
+          fetch-depth: 2
           sparse-checkout-cone-mode: false
           sparse-checkout: |
             /*
@@ -77,6 +78,21 @@ jobs:
 
       - name: Composer dump
         run: composer dump --strict-psr --strict-ambiguous --no-interaction --no-scripts
+
+      - name: Find new PHP files
+        if: github.event_name == 'pull_request'
+        id: new-php-files
+        run: |
+          # HEAD is the PR merge commit, HEAD~1 is its first parent (base branch tip)
+          NEW_PHP_FILES=$(git diff -z --diff-filter=A --name-only HEAD~1 HEAD -- 'app/**.php' 'bootstrap/**.php' 'config/**.php' 'database/**.php' 'LibreNMS/**.php' 'resources/**.php' 'routes/**.php' 'tests/**.php' | xargs -0)
+          echo "Discovered new PHP files: $NEW_PHP_FILES"
+          echo "files=$NEW_PHP_FILES" >> "$GITHUB_OUTPUT"
+
+      - name: Run PHPStan (New files, level 8)
+        if: github.event_name == 'pull_request' && steps.new-php-files.outputs.files != ''
+        env:
+          NEW_PHP_FILES: ${{ steps.new-php-files.outputs.files }}
+        run: ./vendor/bin/phpstan analyze --no-interaction --memory-limit=4G --level=8 -- $NEW_PHP_FILES
 
       - name: Run PHPStan
         run: ./vendor/bin/phpstan analyze --no-interaction --memory-limit=4G

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -88,11 +88,11 @@ jobs:
           echo "Discovered new PHP files: $NEW_PHP_FILES"
           echo "files=$NEW_PHP_FILES" >> "$GITHUB_OUTPUT"
 
-      - name: Run PHPStan (New files, level 8)
+      - name: Run PHPStan (New files, level 6)
         if: github.event_name == 'pull_request' && steps.new-php-files.outputs.files != ''
         env:
           NEW_PHP_FILES: ${{ steps.new-php-files.outputs.files }}
-        run: ./vendor/bin/phpstan analyze --no-interaction --memory-limit=4G --level=8 -- $NEW_PHP_FILES
+        run: ./vendor/bin/phpstan analyze --no-interaction --memory-limit=4G --level=6 -- $NEW_PHP_FILES
 
       - name: Run PHPStan
         run: ./vendor/bin/phpstan analyze --no-interaction --memory-limit=4G

--- a/phpstan.neon
+++ b/phpstan.neon
@@ -40,5 +40,3 @@ parameters:
 
     parallel:
         processTimeout: 300.0
-
-

